feat(IconPicker): upgraded to FontAwesome5 and added IconPicker

bleucitron opened the pull request:
fix(FontAwesome): used pure FontAwesome CDN

feat(IconPicker): listed of FA icons

feat(IconPicker): okay version

feat(IconPicker): added Scrollable

feat(IconPicker): added IconPicker

bleucitron committed:
feat(IconPicker): upgraded to FontAwesome5 and added IconPicker

fix(FontAwesome): used pure FontAwesome CDN

feat(IconPicker): listed of FA icons

feat(IconPicker): okay version

feat(IconPicker): added Scrollable

feat(IconPicker): added IconPicker

GuillaumeAmat commented:
@iOiurson The code coverage has decreased and it is blocking the PR.

GuillaumeAmat commented on package.json:
```diff
@@ -79,16 +79,15 @@
-    "font-awesome": "^4.7.0",
     "hex-rgb": "^1.0.0",
     "jquery": "^3.2.1",
+    "js-yaml": "^3.12.0",
```
The weight of that library is huge: https://bundlephobia.com/result?p=js-yaml@3.12.0

Is there any other option?

bleucitron replied:
[yamljs](https://bundlephobia.com/result?p=yamljs@0.3.0) is the lightest i've found.
Would that be ok ?

The only drawback is i can't make it compile with webpack, i get 
```
Module not found: Error: Can't resolve 'fs' in '/Users/Romain/Code/osm-ui-react/node_modules/yamljs/lib'
```
Seems [doable](https://github.com/jeremyfa/yaml.js/issues/102), but I couldn't figure it out yet...

GuillaumeAmat replied:
I'm afraid we'll have to do it another way... 171 kB is too huge for a low-level module IMO.

As we can see in the `src/components/IconPicker/IconPicker.jsx` file, the FontAwesome version is related to the OSM UI version. I mean, for a build of OSM UI there is a version of FontAwesome.

So what about this:
* We create a script to download the YAML file and convert it into a JSON format
* At the build time of OSM UI, we launch that script
* At the run time, we use the JSON file

That way we have the same level of feature and less weight to carry :)

P.S.: For the launch of the new script, you have the choice of creating a Makefile or using the `pre` prefix in the `package.json` file

```
"prebuild": "download and convert YAML",
"build": "same command as before",
```

bleucitron replied:
The data i need as JSON is 49kB, will that be ok ?

GuillaumeAmat replied:
It's more than 3 times lighter than the yaml module itself, and we don't need to process/convert the yaml file at the runtime 👍

GuillaumeAmat commented on src/components/Loader/Loader.jsx:
```diff
@@ -86,5 +86,6 @@ Loader.defaultProps = {
 };
 
 Loader.displayName = 'components/Loader';
```
Can you fix that `displayName` please? It should just be `Loader`.

GuillaumeAmat commented on src/components/Scrollable/Scrollable.jsx:
```diff
@@ -0,0 +1,112 @@
+    const element = this.scrollable.current;
+
+    if (element && window.getComputedStyle(element).overflowY !== 'hidden') {
+      console.log('scrollHeight', element.scrollHeight);
```
Can you remove all the `console.log` please?

GuillaumeAmat commented on src/components/Scrollable/Scrollable.jsx:
```diff
@@ -0,0 +1,112 @@
+  checkScroll() {
+    const element = this.scrollable.current;
+
+    if (element && window.getComputedStyle(element).overflowY !== 'hidden') {
```
As you use `window`, the `checkScroll` method becomes unpure and not testable. You should use `window` in an helper, which can be mocked during the tests.

bleucitron replied:
I'm not sure i understand...
Is this ok:
- in regular cases, the component uses the helper
- in test env, the component is provided an optional prop mock function that supersedes the helper
?

GuillaumeAmat replied:
Hmm... I'm not sure to understand what you propose :/

What I meant is:
* The `checkScroll` method calls `window` (which is an external dependency as it not provided by an argument)
* Plus, `window` is related to the browser environment so it is not available in tests without `jsdom` for instance

So I proposed to put that call in an external function/helper which can be mocked during the tests.

Eg:

```js
// src/helpers/browser.js
export function isOverflowYHidden(element) {
  return window.getComputedStyle(element).overflowY !== 'hidden';
}
```

And during your tests:

```js
jest.mock('helpers/browser.js', () => {
  isOverflowYHidden: () => true,
})
```

Or `false`, whatever. But the further tests will internally call `isOverflowYHidden` and it will always return the value you set in the mock. The behavior is controlled, the earth is safe 💪

GuillaumeAmat commented on src/components/IconPicker/IconPicker.jsx:
```diff
@@ -0,0 +1,365 @@
+
+  componentDidMount() {
+    fetch(
+      'https://raw.githubusercontent.com/FortAwesome/Font-Awesome/5.2.0/advanced-options/metadata/categories.yml'
```
The URL should be in an imported `const`.

GuillaumeAmat commented on src/components/IconPicker/IconPicker.jsx:
```diff
@@ -0,0 +1,365 @@
+      .catch(e => console.error(`Error in fetching categories ${e}`));
+
+    fetch(
+      'https://raw.githubusercontent.com/FortAwesome/Font-Awesome/5.2.0/advanced-options/metadata/icons.yml'
```
The URL should be in an imported `const`.

bleucitron commented:
@GuillaumeAmat I'm struggling with the async testing...
If i mock the async function that's in my `componentDidMount`, it kind of works, but seems a bit off:
- the snapshot i get is huge (from the Promise values i guess)
- i can't really test the component after it's totally mounted, since i have an async part in the `didMount` process. This could be a [solution](https://stackoverflow.com/questions/38308214/react-enzyme-test-componentdidmount-async-call), but does not seem 100% right...

**I kind of get the feeling that components with async `didMount` are not properly testable.** Maybe we should write sync components that get their initial data as props after a prior fetch has been called.

What do you think ?

GuillaumeAmat commented:
@iOiurson You can inspire you from `osmose-request`.

The call for the data has to be done in an helper/external function and mocked during the test.

```js
componentDidMount(){
  fetchData();
}
```

```js
// src/helpers/myService.js

export function fetchData() {
  // ...
}
```

```js
// src/helpers/__mocks__/myService.js
export const fetchData = jest.fn().mockImplementation(() => ({
  someData: [1, 2, 3],
}));
```

That way, the test will always call the mock and return the same data.

For the huge snapshot thing, I think we have to use enzyme's `shallow` rather than the other methods. That way we unit test the rendered components and not the children.

bleucitron committed:
fix(Scrollable): better test

bleucitron committed:
fix(IconPicker): download data at buildtime and added test

bleucitron commented:
Rebasing my other branches from develop, I now get:
`Can't resolve '../node_modules/bootstrap/less/bootstrap' in '/Users/Romain/Code/osm-ui-react/src'`

I suppose #60 should resolve this, but if we are not merging it right now, do you have any idea how i could solve this issue ?

GuillaumeAmat commented:
Indeed, #60 will kick its ass ;)

Are you sure that the path exists? Is the bootstrap module installed?

GuillaumeAmat commented:
I just tried to pull the branch and then:
* `rm -rf node_modules`
* `npm install`
* `npm run storybook`

And it went just fine, no error in the terminal nor the browser.

bleucitron commented:
For some reason, in my case, in `src/bootstrap.less`

```
@import '../node_modules/bootstrap/less/bootstrap';
```

fails, whereas
```
@import '../node_modules/bootstrap/less/bootstrap.less';
```
works.

I have no idea why this did not occur before, and occurs now...

bleucitron committed:
fix(IconPicker): adjusted tests

bleucitron committed:
fix(IconPicker): verify onChoose props exists before using it

GuillaumeAmat commented:
@iOiurson You were right, there's still something to do here:

* https://github.com/osm-ui/react/pull/57#discussion_r207747636
* https://github.com/osm-ui/react/pull/57#issuecomment-414442276

But I really need to go forward instead of strugling with Travis & co, so I tried to import the project to Gitlab and it worked great, as that PR is now an MR in Gitlab: https://gitlab.com/osm-ui/react/merge_requests/57

Can you consider finishing your work there?

If so, thanks dude! 🤙

bleucitron commented:
Actually the 2 points you're talking about are resolved:
- i'm using [yamljs](https://bundlephobia.com/result?p=yamljs@0.3.0) instead of [js-yaml](https://bundlephobia.com/result?p=js-yaml@3.12.0), which is way lighter. I could not make it work before with webpack, but since now i'm only using it in Node it's ok.
- i don't have a `didMount` anymore since i'm requiring pre-loaded json files

In my opinion, all is done (apart from the recent conflicts).
Can you recheck, i will solve the conflicts ?

GuillaumeAmat commented:
> Actually the 2 points you're talking about are resolved:

Oh ok cool then, but...

> i'm using yamljs instead of js-yaml, which is way lighter. I could not make it work before with webpack, but since now i'm only using it in Node it's ok.

The YAML file is still downloaded, parsed and converted in the browser while we could convert it at the build time and just consume a plain JSON in the browser. IMHO (almost) every CPU waste must be avoided for low-power devices sake.

bleucitron commented:
> The YAML file is still downloaded, parsed and converted in the browser

Is it ? I added the `prebuild` and `prestorybook` scripts to dl, parse and convert the yaml into json, which is imported to the component.

Unless i misunderstand something completely ... i think the current state of the PR does what you expect.

bleucitron commented:
BTW, discussions like these might feel easier in a chat-like tool...
